pkg/sitewise/datasource.go

```diff
@@ -23,13 +25,56 @@ type Datasource struct {
 }
 
 func NewDatasource(settings backend.DataSourceInstanceSettings) (*Datasource, error) {
-	cfg := client.AWSSiteWiseDataSourceSetting{}
+	cfg := models.AWSSiteWiseDataSourceSetting{}
+
 	err := cfg.Load(settings)
 	if err != nil {
 		return nil, err
 	}
-	sessions := awsds.NewSessionCache()
 
+	err = cfg.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Region == models.EDGE_REGION && cfg.EdgeAuthMode != models.EDGE_AUTH_MODE_DEFAULT {
+		// TODO: refresh session every 4h since creds expire
+		edgeAuthenticator := EdgeAuthenticator{
+			Settings: cfg,
+		}
+
+		var waitTime time.Duration
+
+		updateAuth := func() error {
+			authInfo, err := edgeAuthenticator.Authorize()
+			if err == nil {
+				cfg.AccessKey = authInfo.AccessKeyId
```

Comment:
Need to use mutex since reading and writing happens in differrent goroutines, very very rare of course, but :)

Reply:
```go
type Config struct {
	Data string
}

func NewClient(cfg Config) {
	fmt.Println(cfg.Data)
}

func TestConcurrentReadWrite(t *testing.T) {
	cfg := Config{
		Data: "test",
	}
	//var mu sync.Mutex
	done := make(chan struct{})
	go func() {
		//mu.Lock()
		cfg.Data = "1"
		//mu.Unlock()
		close(done)
	}()
	//mu.Lock()
	cfgCopy := cfg
	//mu.Unlock()
	NewClient(cfgCopy)
	<-done
}
```

If you run this (which is an close analogue I believe) with race detector:

```
go test -race
```

Then race detector will report data races. With mutex code uncommented race detector is happy.

Reply:
thanks for the catch!